### PR TITLE
Buy clout with eth

### PR DESCRIPTION
**Pull request by maebeam**

### CODE REVIEW

File: src/app/crypto.service.ts [hunk 1]
```diff
@@ -172,4 +173,38 @@ export class CryptoService {
 
     return bs58check.encode(prefixAndKey);
   }
+
+  // NOTE: Our ETH address uses the bitcion/bitclout derivation path, not the ETH path.
```
Comment (redpartyhat) [inline src/app/crypto.service.ts:177]: nit: bitcoin*

File: src/app/crypto.service.ts [hunk 2]
```diff
@@ -172,4 +173,38 @@ export class CryptoService {
...
+  //
+  // Reference implementation: https://github.com/ethereumjs/ethereumjs-util/blob/master/src/account.ts#L249
+  keychainToEthAddress(keychain: HDNode, network: Network): string {
+    // Get the uncompressed key
+    const ec = new EC('secp256k1');
+    const publicKeyEC = ec.keyFromPublic(keychain.publicKey, 'array');
+
+    // ETH uses the last 40 characters of the 64 byte SHA3 Keccak 256
+    const uncompressedKey = Buffer.from(publicKeyEC.getPublic(false, 'array').slice(1));
+    const ethAddress = new Keccak(256).update(uncompressedKey).digest('hex').slice(24);
```
Comment (redpartyhat) [inline src/app/crypto.service.ts:194]: In the reference implementation you've linked to in the comment for this function, it says:
```
  // Only take the lower 160bits of the hash
  return keccak(pubKey).slice(-20)
```
^This code only takes 20 bytes while yours takes 40.  Am I reading that correctly? Can you explain what the difference is?

File: src/app/crypto.service.ts [hunk 3]
```diff
@@ -172,4 +173,38 @@ export class CryptoService {
...
+  // NOTE: Our ETH address uses the bitcion/bitclout derivation path, not the ETH path.
+  // This is ugly but we only do it because they're just deposit addresses and we couldn't
+  // backfill this data for existing users because we store the derived private key.
+  // A user can still easily import their account to an ETH client to recover any ETH
+  // by coping the seedHex from local storage.
+  //
+  // We aren't using ethereumjs-util to minimize bundle size and protect from vulnerabilities
+  // until LavaMoat is ready
+  //
+  // Reference implementation: https://github.com/ethereumjs/ethereumjs-util/blob/master/src/account.ts#L249
```
Comment (redpartyhat) [inline src/app/crypto.service.ts:186]: Thanks for linking to this reference implementation.  Super helpful!

File: src/app/crypto.service.ts [hunk 4]
```diff
@@ -172,4 +173,38 @@ export class CryptoService {
 
     return bs58check.encode(prefixAndKey);
   }
+
+  // NOTE: Our ETH address uses the bitcion/bitclout derivation path, not the ETH path.
+  // This is ugly but we only do it because they're just deposit addresses and we couldn't
+  // backfill this data for existing users because we store the derived private key.
+  // A user can still easily import their account to an ETH client to recover any ETH
+  // by coping the seedHex from local storage.
```
Comment (redpartyhat) [inline src/app/crypto.service.ts:181]: Have you confirmed that this works?

Comment (redpartyhat) [inline src/app/crypto.service.ts:181]: Have we checked that this works?  Like that the seedhex in identity perfectly maps to the eth address that this function produces?

Comment (maebeam) [reply src/app/crypto.service.ts:181]: Yep, verified

File: src/app/crypto.service.ts (hunk 2 above)
Comment (maebeam) [reply src/app/crypto.service.ts:194]: 40 hex characters is 20 bytes :)

Comment (redpartyhat) [reply src/app/crypto.service.ts:194]: Great, makes sense!

File: src/app/crypto.service.ts (hunk 4 above)
Comment (redpartyhat) [reply src/app/crypto.service.ts:181]: Okay dope.  I feel pretty good about the addresses then :)

File: src/app/app.component.ts [hunk 5]
```diff
@@ -54,6 +52,9 @@ export class AppComponent implements OnInit {
       this.globalVars.hideGoogle = true;
     }
 
+    // Migrate all accounts
+    this.accountService.migrate();
```
Comment (redpartyhat) [inline src/app/app.component.ts:56]: Migrate seems dope, love it.

File: src/app/account.service.ts [hunk 6]
```diff
@@ -71,26 +73,31 @@ export class AccountService {
...
-
-    localStorage.setItem(AccountService.usersStorageKey, JSON.stringify(privateUsers));
-
-    return publicKey;
+  // Public Modifiers
+
+  addUser(keychain: HDKey, mnemonic: string, extraText: string, network: Network, google?: boolean): string {
+    const seedHex = this.cryptoService.keychainToSeedHex(keychain);
+    const btcDepositAddress = this.cryptoService.keychainToBtcAddress(keychain, network);
+    const ethDepositAddress = this.cryptoService.keychainToEthAddress(keychain);
```
Comment (diamondhands0) [inline src/app/account.service.ts:81]: It's fine, but curious: Is there a reason we moved the pubkey generation for BTC/ETH to identity from backend where it was before?

File: src/app/crypto.service.ts (hunk 3 above)
Comment (diamondhands0) [reply src/app/crypto.service.ts:186]: +1

**Pull request merged**